[adispezio][cfierro] Add widget to module migration script.

ajperel opened the pull request:
If a widget and a module with the same widget and it's patterns are in the same book, this script can update all the content to change references from the non-modular widget to the modular one, and delete the widget and associated non-modular patterns. Expected flow is:

(1) Someone creates a modular version of widget + patterns
(2) Module is uploaded to Module Repo
(3) Module is installed into projects from Repo / UI
(4) This script is run to cleanup all projects that now have two forms of the widget & patterns.

This script adds a repo dependency on the Beautiful Soup 4 HTML parsing library.

ajperel commented:
If you want to test this script for yourself you can make a csv spec file with the contents:

```
tsn_250b, testing, annotated_image, inkling.annotated_image
tsn_250b, testing, flashcard, inkling.flashcard
tsn_250b, testing, multiple_choice, inkling.multiple_choice
tsn_250b, testing, show_hide, inkling.show_hide
tsn_250b, testing, slideline, inkling.slideline
tsn_250b, testing, slideshow, inkling.slideshow
tsn_250b, testing, timeline, inkling.slideshow
```

and run

`migrate.py -c <your csv config file> -s`

The `-s` flag will stop the script from actually doing an SVN commit, keeping the project clean for testing the migration script. This is the migration project I shared with all of you. Feel free to try on other projects too. If you do commit the changes / omit the flag (which I have done at least once to test the whole flow) just rollback the SVN changes.

Before merging I will work with Anthony to test this on a few more projects too.

cfierro commented:
Sorry for the delay. Starting review on this.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+            try:
+                for row in reader:
+                    # Strip any whitespace in row contents.
+                    row = map(str.strip, row)
```
nit: It is generally more pythonic to use list comprehension instead of map. Something like `row = [column.strip() for column in row]` would also help get the message across that row is a list of column values. For a sec, I thought row was 'col1,col2,col3'.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+                    # Strip any whitespace in row contents.
+                    row = map(str.strip, row)
+                    if len(row) < 4 or not (row[0] and row[1] and row[2] and
+                            row[3]):
```
We can probably simplify this to use `all`. `if len(row) < 4 or not all(row)`. Then we can just bump the length check if we ever need to add more columns.

ajperel replied:
I didn't know about `all`. Awesome!

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+    configuration file.
+    """
+    results = []
+    if args.config:
```
I believe you can get rid of this check. It appears you made it a required argument above.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+    relative to the non-modular widget to be relative to the modular widget.
+    Also fixes any JSON configuration files references from a widget's <object>
+    tag.
+    """
```
Nit: missing args section.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+    """
+    log.debug('Reading HTML file: %s', filePath)
+    with open(filePath, 'rb') as htmlFile:
+        lines = ''.join(htmlFile.readlines())
```
Since we are joining all the lines, it might be clearer to rename `lines` to `htmlContent`. Makes it a bit clearer below that it isn't an array of strings.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+                # exists. If it does, assume it is a relative path to something
+                # in the project and update it to be relative to the new modular
+                # widget location.
+                val = param['value']
```
Param tags are always guaranteed to have value attributes right? I wouldn't be opposed of using `param.get('value')` instead in case someone has a really odd param tag without a value.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+        # leaves all parameters and JSON config files incorrectly relative to
+        # the wrong location.
+        lines = re.sub(r'(<object\s.*?data=")' + widgetViewRelativePath +'"',
+                       r'\g<1>' + relativeModuleViewPath + '"',
```
In the regex below you extracted the first argument to a variable called `toReplace`. I found that really helpful. Could you also extract the second argument to another variable called maybe `replacement`.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+                    toReplace = (r'(<object.*?data="' + relativeModuleViewPath +
+                        '".*?<param.*?value=")' + val + '(".*?)(?=</object>)')
+                    lines = re.sub(toReplace,
+                                   r'\g<1>' + newPath + r'\g<2>',
```
Same comment about extracting this regex to a variable named `replacement` for readability.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+
+                    # If the param is for the widget config file, also clean up
+                    # that config file.
+                    if param['name'] == 'configFile':
```
Same comment about using `param.get('name')` instead.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+        # Write out all the changes.
+        log.info('Writing updated HTML file: %s', filePath)
+        with open(filePath, 'wb') as htmlFile:
+           htmlFile.write(lines)
```
Kind of odd to do a `with open` within another `with open` on the same file. We can unindent 192-195 once and it should still work.

ajperel replied:
Whoops. Great call.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+            modularWidthPath - String absolute path to the modular widget.
+    """
+    log.debug('Reading JSON config file: %s', filePath)
+    with open(filePath, 'rb') as configFile:
```
`_updateHTMLFile` is wrapped within a try/except, so if this throws an IOError, things should be okay. Do we, however, want to try catching bad config file paths here so that the rest of the changes still get written?

ajperel replied:
Very torn on this. It's kind of nice, but you're still in the exact same state where you have to manually fix the migration somewhere (config file is still wrong). I'd almost just wait and see how often it happens. I expect if they are problems they are far more likely to be that the scripts do the wrong thing, than that they have file i/o errors. And if we don't just let it propagate, how do we nicely report up the errors. Thoughts?

cfierro replied:
I'm torn on it too, so let's leave it as is for now.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+    """
+    log.debug('Reading JSON config file: %s', filePath)
+    with open(filePath, 'rb') as configFile:
+        lines = ''.join(configFile.readlines())
```
Same comment about renaming `lines`. Maybe to `jsonString` here instead.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+    log.debug('Reading HTML file: %s', filePath)
+    with open(filePath, 'rb') as htmlFile:
+        lines = ''.join(htmlFile.readlines())
+        soup = BeautifulSoup(lines)
```
I like how in `_updateConfigFile` you made your `with open` block as small as possible (just read the lines). Could you also do that here? That way we can unindent a lot the code below.

cfierro commented on modules/migrate.py:
```diff
@@ -0,0 +1,485 @@
+    if os.path.exists(patternFilePath):
+        log.debug('Reading pattern snippet file: %s', patternFilePath)
+        with open(patternFilePath) as patternFile:
+            lines = ''.join(patternFile.readlines())
```
Same comment about renaming.

cfierro commented:
Done with comments.

ajperel commented:
@cfierro - Addressed all comments and learned a bit about Python. Thanks for the good review! Torn on improving the error handling when writing config files. I think we could make some minor improvements but I'm not sure how important it is given that there will always be that error case.

cfierro commented:
LGTM
